blog/fb4/index.qmd

```diff
@@ -0,0 +1,162 @@
+                                comparators=[
+                                    Name(id='c', ctx=Load())]),
+                            body=[
+                                Return(
+                                    value=Constant(value='Isosceles'))],
+                            orelse=[
+                                If(
+                                    test=Compare(
+                                        left=Name(id='a', ctx=Load()),
+                                        ops=[
+                                            Eq()],
+                                        comparators=[
+                                            Name(id='c', ctx=Load())]),
+                                    body=[
+                                        Return(
+                                            value=Constant(value='Isosceles'))],
+                                    orelse=[
+                                        Return(
+                                            value=Constant(value='Scalene'))])])])],
+            decorator_list=[])],
+    type_ignores=[])
+```
+
+A simple way to produce a valid mutated version of a program is through replacing some of its statements with ‘pass’. While MuFunctionAnalyzer utilizes mutations of specific functions, MuProgramAnalyzer is the main class responsible for mutation analysis of the test suite, accepting the name of the module to be tested and its source code. The method ‘getitem’ accepts the test module, fixes the import entries on the test module to correctly point to the mutant module, and passes it to the test runner ‘MutantTestRunner’. The problem lies with equivalent mutants, mutants indistinguishable from the original in terms of semantics, whereas it becomes very difficult to judge the mutation score in their presence. A solution to this is inspecting mutants manually if they’re small enough, or randomly selecting a smaller amount of mutants to manually inspect if there’s a larger scale of them. Chao’s estimator can also be utilized to compute the result of the complete test matrix of each test against each mutant.
+
+## Reflection
+
+The Fuzzing Book has a chapter that talks about something called mutation analysis. This is a way to check how good our tests are when we run software. Just because we test a lot doesn't mean the software will always work right. The book gives two examples, named ineffective_test_1 and ineffective_test_2, to show this point.
+
+In the same chapter, there's also a bit about a triangle program. This program figures out the type of triangle using the lengths of its three sides. To check how good the tests are for this program, they first make the program more standard. They do this by using a method called 'mutation'. After that, they change some parts of the program, like swapping 'return' with 'pass'. If the program works correctly, it should show an error because some triangle types won't be recognized.
```

Comment:
Why is this content about the triangle program relevant here?

Reply:
I added the relevance of the triangle program. In summary I went in depth about the importance of the triangle program and how it pertains to mutation.